ext/liquid_c/vm.c

```diff
@@ -11,53 +12,394 @@ ID id_ivar_last_capture_length;
                 size_t size = *const_ptr++;
                 rb_str_cat(output, text, size);
+                increment_write_score(args);
                 break;
             }
             case OP_WRITE_NODE:
-                rb_funcall(cLiquidBlockBody, id_render_node, 3, context, output, (VALUE)*const_ptr++);
-                if (RARRAY_LEN(interrupts)) {
-                    return;
+                rb_funcall(cLiquidBlockBody, id_render_node, 3, args->context, output, (VALUE)*const_ptr++);
+                if (RARRAY_LEN(vm->interrupts)) {
+                    return false;
                 }
+                increment_write_score(args);
+                break;
+            case OP_RENDER_VARIABLE_RESCUE:
+                args->node_line_number = (unsigned int)*const_ptr++;
+                args->ip = ip;
+                args->const_ptr = const_ptr;
+                break;
+            case OP_POP_WRITE_VARIABLE:
+            {
+                VALUE var_result = vm_stack_pop(vm);
+                if (vm->global_filter != Qnil)
+                    var_result = rb_funcall(vm->global_filter, id_call, 1, var_result);
+                write_obj(output, var_result);
+                args->ip = NULL;
+                increment_write_score(args);
+                break;
+            }
+            case OP_PUSH_CONST:
+                vm_stack_push(vm, (VALUE)*const_ptr++);
+                break;
+            case OP_HASH_NEW:
+            {
+                size_t hash_size = *ip++;
+                size_t num_keys_and_values = hash_size * 2;
+                VALUE hash = rb_hash_new();
+                VALUE *args_ptr = vm_stack_pop_n_use_in_place(vm, num_keys_and_values);
+                hash_bulk_insert(num_keys_and_values, args_ptr, hash);
+                vm_stack_push(vm, hash);
+                break;
+            }
+            case OP_FILTER:
+            {
+                VALUE filter_name = (VALUE)*const_ptr++;
+                uint8_t num_args = *ip++; // includes input argument
+                VALUE *args_ptr = vm_stack_pop_n_use_in_place(vm, num_args);
+                VALUE result = vm_invoke_filter(vm, filter_name, num_args, args_ptr);
+                vm_stack_push(vm, result);
                 break;
+            }
+            case OP_PUSH_EVAL_EXPR:
+            {
+                VALUE expression = (VALUE)*const_ptr++;
+                vm_stack_push(vm, context_evaluate(args->context, expression));
+                break;
+            }
+
             default:
                 rb_bug("invalid opcode: %u", ip[-1]);
         }
-        if (RB_UNLIKELY(is_captured)) {
-            rb_funcall(resource_limits, id_increment_write_score, 1, output);
-        } else if (RSTRING_LEN(output) > render_length_limit) {
-            rb_funcall(resource_limits, id_raise_limits_reached, 0);
-        }
+    }
+}
+
+void liquid_vm_next_instruction(const uint8_t **ip_ptr, const size_t **const_ptr_ptr)
```

Comment:
These constants are pointer sized operands to the instruction, similar to one the byte sized ones.

In hindsight, the instruction pointer should be been a struct with a pointer to both to provide a bit of a zero-cost abstraction around them.  We are also missing an abstraction for accessing those operands from the block body code.  However, this PR is at least a step in the right directly by moving the vm assembler out of block.c.

However, all of this is still internal to liquid-c, so we can definitely change the VM instructions.

Reply:
It's not really an operand if it's not in the bytecode, in an instruction. So it's a bit like having two sets of instructions. The problem we'll have is keeping them in sync all the time if we modify the bytecode.

Not sure what you mean RE the struct with a pointer.

But all good, you're right, step in the right direction!

Reply:
Yes, that is what I meant for the instruction pointer struct.

I want the constants separate, because they will require serialization and deserialization, unlike the byte code which is already serialized and doesn't require deserialization for execution.  Ruby's in-memory VM code interleaves the pointers immediately into the instruction, which ended up leading to pointer size instructions (to avoid alignment issues) and requires `RubyVM::InstructionSequence.load_from_binary` to deserialize the instructions.

Of course, we can also have a constant table, which is ideal for constants that get re-used (e.g. filter names, variable names and object keys are common to re-use) and allows only those ruby constants to be deserialized.  I want to switch to that after adding serialization support.